### DISCUSSION TITLE
vm.startBroadcast

### DISCUSSION THREAD

**Discussion by JanSzczepan** [Cyfrin/foundry-full-course-cu#513] [Q&A]
Hey,
Can someone explain to me vm.startBroadcast and vm.stopBroadcast functions? Probably I don't really understand them, for example I dont know why we need them in HelperConfig belowe...

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {Script, console} from "forge-std/Script.sol";
import {VRFCoordinatorV2Mock} from "../test/mock/VRFCoordinatorV2Mock.sol";

contract HelperConfig is Script {
   struct NetworkConfig {
      address vrfCoordinatorV2Address;
      bytes32 gasLane;
      uint64 subscriptionId;
      uint32 callbackGasLimit;
      uint256 entraceFee;
      uint256 interval;
   }

   NetworkConfig public activeNetworkConfig;

   event HelperConfig__CreatedMockVRFCoordinator(address vrfCoordinator);

   constructor() {
      if (block.chainid == 11155111) {
         activeNetworkConfig = getSepoliaNetworkConfig();
      } else {
         activeNetworkConfig = getOrCreateAnvilNetworkConfig();
      }
   }

   function getOrCreateAnvilNetworkConfig() public returns (NetworkConfig memory anvilNetworkConfig) {
      if (activeNetworkConfig.vrfCoordinatorV2Address != address(0)) {
         return activeNetworkConfig;
      }

      uint96 baseFee = 0.25 ether;
      uint96 gasPriceLink = 1e9;

      vm.startBroadcast();
      VRFCoordinatorV2Mock vrfCoordinatoV2Mock = new VRFCoordinatorV2Mock(baseFee, gasPriceLink);
      vm.stopBroadcast();

      emit HelperConfig__CreatedMockVRFCoordinator(address(vrfCoordinatoV2Mock));

      anvilNetworkConfig = NetworkConfig({
         vrfCoordinatorV2Address: address(vrfCoordinatoV2Mock),
         gasLane: 0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c,
         subscriptionId: 0,
         callbackGasLimit: 500000,
         entraceFee: 0.1 ether,
         interval: 60
      });
   }

   function getSepoliaNetworkConfig() public pure returns (NetworkConfig memory sepoliaNetworkConfig) {
      sepoliaNetworkConfig = NetworkConfig({
         vrfCoordinatorV2Address: 0x8103B0A8A00be2DDC778e6e7eaa21791Cd364625,
         gasLane: 0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c,
         subscriptionId: 0,
         callbackGasLimit: 500000,
         entraceFee: 0.1 ether,
         interval: 60
      });
   }
}

```

**Comment by shikhar229169** [comment]
I have already answered it in discussion #504, please refer to that.

**Comment by JanSzczepan** [comment]
Ok, thank you I missed this one.
So now I understand that transactions between vm.startBroadcast and vm.stopBroadcast are called on behave of account with private key that we specify. 
Is there any other difference?

**Comment by JanSzczepan** [comment]
And to be honest I still don't understand why we need startBroadcast in above HelperConfig script. When I remove this two lines (vm.startBroadcast and vm.stopBroadcast) it works fine

**Reply by shikhar229169** [reply]
The thing is that if you don't add broadcast, only simulation happens, no doubt address would be generated but the contract deployment is just a simulation and it is actually not deployed on the network.

If you run command to deploy your contract to anvil, it would for sure deploy your contract, but the vrf coordinator mock will not get deployed on the anvil network (only simulation happens), and if you try to interact with your coordinator it will not work because the contract is not available on the network.

So, we need to add broadcast only for those txns which we want to send to network.
But if you don't add broadcast it will not be sent to the network, only simulation takes place.

**Reply by 0xSimeon** [reply]
So in that context, how does `vm.startBroadcast() differs from `vm.startPrank(user we want to prank as); `

From what I know: 
- `vm.startBroadcast();` just picks the foundry default key and simulate a tx. 
- `vm.startPrank(USER); ` does a tx as well but with the specific user we pass in as argument to the `startPrank() function` 

I will appreciate further clarification on this.

**Reply by shikhar229169** [reply]
Prank is used for tests. Its just only for testing your contracts, and a adding a prank will never do a actual txn on chain (its all just a simulation). 
And to perform real txn on chain you need to add broadcast. Another thing is prank and broadcast don't work together.

Now consider the below command:
```bash
forge script script/deployFundMe.s.sol --rpc-url $RPC_URL --private-key $ANVIL_PRIVATE_KEY --broadcast
```

Its just a simple command which deploys the contract FundMe on anvil chain and notice that we passed private key as argument, so where ever we add vm.startBroadcast(), the txn would be performed on behalf of the user whose private key is passed as argument and as we are using vm.startBroadcast it will be sent as actual txn on the network.
